Document package design choices

Kolaru opened the pull request:
As discussed last Friday, I have drafted an addition to the documentation to discuss some design choices that we have  made.

I have summarized all discussion that I could think of for now, but please let me know if any other controversies should be mentionned to.

OlivierHnt commented on docs/src/philosophy.md:
```diff
@@ -0,0 +1,127 @@
+the resulting intervals are guaranteed to bound the true image of the starting
+intervals.
+
+Due to the way that the julia ecosystem has evolved,
+this means that `Interval` **must** be a subtype of `Real`,
+as it is the default supertype used to describe
+"numbers that are not complex".
```
Maybe "must" is too strong?

"
Due to the evolution of the Julia ecosystem -- specifically that `Real` serves as the default supertype for numerical types that are not complex --, it is particularly advantageous for `Interval` to be defined as a subtype of `Real`.
"

Kolaru replied:
I have tuned it down a little bit, but I like "must" so that it is clear it is not open for discussion (at least for the current document).

OlivierHnt commented on docs/src/philosophy.md:
````diff
@@ -0,0 +1,127 @@
+(note that it holds for **all** real numbers in `X`,
+even those that can not be represented as floating point numbers):
+```julia
+f(x) ∈ f(X)
````
Should we use a math-block instead? Since `Base.in` is ambiguous; e.g.,

```math
f(x) \in f(X), \qquad \forall x \in X
```

OlivierHnt commented on docs/src/philosophy.md:
````diff
@@ -0,0 +1,127 @@
+With this, we want to have, for any function`f`,
+for all `x` in `X` and all `y` in `Y`,
+```julia
+f(x, y) ∈ f(X, Y)
````
Same comment, maybe
```math
f(x, y) \in f(X, Y), \qquad \forall x \in X, \, y \in Y
```

OlivierHnt commented on docs/src/philosophy.md:
```diff
@@ -0,0 +1,127 @@
+In this way, we can return a more informative error,
+but we only do it when the result is ambiguous.
+
+This has a clear cost, however, in that some expected behaviors do not hold.
+For example, an `Interval` is not equal to itself.
```
Maybe this can be removed. What you wrote above, seems already very clear.

Kolaru replied:
I think it doesn't hurt to have an example.

OlivierHnt commented on docs/src/philosophy.md:
```diff
@@ -0,0 +1,127 @@
+
+## Intervals as sets
+
+We have taken the perspective to always let `Interval`s act as if they were numbers.
```
If we want to not avoid the word "number":

"
We have taken the perspective to always let `Interval`s behave as numerical values.
"

OlivierHnt commented:
Thanks @Kolaru for the PR. It looks really good.

Closes #647.

Maybe due to #675, we can add a sentence (near where `Base.:(==)` is discussed) about `Base.isapprox`?

Kolaru commented:
I've added a summary of the behaviors, I think it address your concerns.

I have also added `setdiff_interval` and `union_interval` for consistency (it simplified the summary :D).

I can add the doc for `@exact`.

I am a bit worried, as the documentation here does not match the behavior on master. Should I port the change from #698 for consistency ?
